types/defines/rpc.d.ts

```diff
@@ -0,0 +1,192 @@
+  }
+  export interface WorkerEntrypointBranded {
+    [__WORKER_ENTRYPOINT_BRAND]: never;
+  }
+  export interface DurableObjectBranded {
+    [__DURABLE_OBJECT_BRAND]: never;
+  }
+  export type EntrypointBranded =
+    | WorkerEntrypointBranded
+    | DurableObjectBranded;
+
+  // Types that can be passed by value over RPC
+  type Serializable<With = never> =
+    | With
+    // Structured cloneables
+    | void
+    | undefined
+    | null
+    | boolean
+    | number
+    | bigint
+    | string
+    | TypedArray
+    | ArrayBuffer
+    | DataView
+    | Date
+    | Error
+    | RegExp
+    // Structured cloneable composites
+    | Map<Serializable<With>, Serializable<With>>
+    | Set<Serializable<With>>
+    | ReadonlyArray<Serializable<With>>
+    | { [key: string | number]: Serializable<With> }
+    // Special types
+    | ReadableStream<Uint8Array>
+    | WritableStream<Uint8Array>
+    | Request
+    | Response
+    | Headers
+    | Stub<Stubable>;
+
+  // Types that can be used through `Stub`s
+  export type Stubable = RpcTargetBranded | ((...args: any[]) => any);
+
+  // Base type for all RPC stubs, including common memory management methods.
+  // `T` is used as a marker type for unwrapping `Stub`s later.
+  interface StubBase<T extends Stubable> extends Disposable {
+    [__RPC_STUB_BRAND]: T;
+    dup(): this;
+  }
+  export type Stub<T extends Stubable> = Provider<T> & StubBase<T>;
+
+  // Recursively rewrite all `Stubable` types with `Stub`s
+  // prettier-ignore
+  type Stubify<T> =
+      T extends Stubable ? Stub<T>
+    : T extends Map<infer K, infer V> ? Map<Stubify<K>, Stubify<V>>
+    : T extends Set<infer V> ? Set<Stubify<V>>
+    : T extends Array<infer V> ? Array<Stubify<V>>
+    : T extends ReadonlyArray<infer V> ? ReadonlyArray<Stubify<V>>
+    : T extends object ? { [K in keyof T]: Stubify<T[K]> }
+    : T;
+
+  // Utility type for adding `Provider`/`Disposable`s to `object` types only.
+  // Note `unknown & T` is equivalent to `T`.
+  type MaybeProvider<T> = T extends object ? Provider<T> : unknown;
+  type MaybeDisposable<T> = T extends object ? Disposable : unknown;
+
+  // Type for method return or property on an RPC interface.
+  // - Serializable types are passed by value.
+  // - Stubable types are replaced by stubs.
+  // - Serializable types containing stubable types are passed by value,
+  //   with stubable types replaced by stubs and a top-level `Disposer`.
+  // Everything else can't be passed over PRC.
+  // Technically, we use custom thenables here, but they quack like `Promise`s.
+  // Intersecting with `(Maybe)Provider` allows pipelining.
+  // For objects, use stub `Promise`s, intersected `Provider`s for pipelining.
+  // prettier-ignore
+  type Result<R> =
+      R extends Serializable ? Promise<R & MaybeDisposable<R>> & MaybeProvider<R>
+    : R extends Stubable ? Promise<Stub<R>> & Provider<R>
+    : R extends Serializable<Stubable> ? Promise<Stubify<R> & Disposable> & MaybeProvider<R>
```

Comment:
The disposer is added regardless of whether there are any stubs in the result. So I think the earlier `R extends Serializable` line should go away, and this line should cover it?